contracts/strategies/dynamic/DynamicLPStrategy.sol

```diff
@@ -0,0 +1,227 @@
+    mapping(address => bool) public strategyExecutors; /// @dev EOAs that can execute safeHarvest
+
+    IDynamicSubLPStrategy[] public subStrategies;
+    IDynamicSubLPStrategy public currentSubStrategy;
+
+    bool public exited; /// @dev After bentobox 'exits' the strategy harvest, skim and withdraw functions can no loner be called
+
+    event LogSubStrategyAdded(address indexed subStrategy);
+    event LogSubStrategyChanged(
+        address indexed fromStrategy,
+        address indexed toStrategy,
+        uint256 amountOut,
+        uint256 amountOutPrice,
+        uint256 amountIn,
+        uint256 amountInPrice
+    );
+    event LogSetStrategyExecutor(address indexed executor, bool allowed);
+
+    /** @param _strategyToken Address of the underlying LP token the strategy invests.
+        @param _bentoBox BentoBox address.
+        @param _strategyExecutor an EOA that will execute the safeHarvest function.
+    */
+    constructor(
+        address _strategyToken,
+        address _bentoBox,
+        address _strategyExecutor
+    ) {
+        strategyToken = _strategyToken;
+        token0 = ISushiSwap(_strategyToken).token0();
+        token1 = ISushiSwap(_strategyToken).token1();
+
+        bentoBox = _bentoBox;
+
+        if (_strategyExecutor != address(0)) {
+            strategyExecutors[_strategyExecutor] = true;
+            emit LogSetStrategyExecutor(_strategyExecutor, true);
+        }
+    }
+
+    modifier isActive() {
+        require(!exited, "BentoBox Strategy: exited");
+        _;
+    }
+
+    modifier onlyBentoBox() {
+        require(msg.sender == bentoBox, "BentoBox Strategy: only BentoBox");
+        _;
+    }
+
+    /// @notice Ensure the current strategy is handling _strategyToken token so that skim,
+    /// withdraw and exit can report correctly back to bentobox.
+    modifier onlyValidStrategy() {
+        require(address(currentSubStrategy) != address(0), "zero address");
+        require(currentSubStrategy.strategyTokenIn() == strategyToken, "not handling strategyToken");
+        _;
+    }
+
+    modifier onlyExecutor() {
+        require(strategyExecutors[msg.sender], "BentoBox Strategy: only Executors");
+        _;
+    }
+
+    function addSubStrategy(IDynamicSubLPStrategy subStrategy) public onlyOwner {
+        require(address(subStrategy) != address(0), "zero address");
+        require(subStrategy.dynamicStrategy() == address(this), "dynamicStrategy mismatch");
+
+        /// @dev make sure the strategy pair token is using the same token0 and token1
+        ISushiSwap sushiPair = ISushiSwap(subStrategy.strategyTokenIn());
+        require(sushiPair.token0() == token0 && sushiPair.token1() == token1, "incompatible tokens");
+
+        subStrategies.push(subStrategy);
+        emit LogSubStrategyAdded(address(subStrategy));
+
+        if (address(currentSubStrategy) == address(0)) {
+            currentSubStrategy = subStrategy;
```

Comment:
There seems to be no check that the first strategy uses the LP we put in?

Reply:
true, I'm adding a require for it, to avoid mistakes for future deployments